### PR TITLE
[ADDED] system model command

### LINKED ISSUE #8176: [mesheryctl] Add mesheryctl system model
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Implementation
1. Enhance `mesheryctl` to expose component generator and component updater functionality in binary form and available in new `mesheryctl` releases.
1. Use new MeshKit libraries. 
1. Create `mesheryctl system model`
    1. Create `mesheryctl system model list` - Use `/api/meshmodels/models` to retrieve and display list of the names of all registered models, using a single column format with just "model" name. 
    1. Create `mesheryctl system model view [model-name]` 
        1. Reusing the logic from `mesheryctl perf` retrieve and display the contents of the model record matching the [model-name]. If multiple models match the name, display list of matches and ask the user to select one for display. 
        1. In a key:value format (one key per line), display all model information, using yaml as the default style of output.
     1. Create `mesheryctl system model update` - supporting the following parameters:
**Usage:**

```
    mesheryctl system model update --source [path-to-spreadsheet] [--output-dir] [relative path persist assets/icons] 
```

**Example:**

```
$ mesheryctl system model update --source https://docs.google.com/spreadsheets/d/e/2PACX-1vSgOXuiqbhUgtC9oNbJlz9PYpOEaFVoGNUFMIk4NZciFfQv1ewZg8ahdrWHKI79GkKK9TbmnZx8CqIe/pub\?gid\=0\&single\=true\&output\=csv --system meshery-server --output-dir /docs/assets/images/models
```

_Future:_
     _1. Create `mesheryctl system model gen` - supporting the following parameters:_

#### Mockups

---

#### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by zakisk**
Signed-off-by: Mohammed Zaki <zs84907@gmail.com>

**Notes for Reviewers**

This PR fixes #8176

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
- [x] Yes, I added unit tests.

### CODE REVIEW

**Review by leecalcote — CHANGES_REQUESTED** [review]
@zakisk will you please see that each error code has a probable cause and a suggested remediation defined?

File: mesheryctl/internal/cli/root/system/error.go
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/error.go:177]: yaml, not json, yeah?

File: mesheryctl/internal/cli/root/system/error.go
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/error.go:173]: @zakisk will you please see that each error code has a probable cause and a suggested remediation defined?

I realize that some or all of these error codes might not be originally defined by you, but it's time to put our foot down and get on the right side of these, making them truly useful, @abdullah1308.

File: mesheryctl/internal/cli/root/system/error.go
Comment (zakisk) [reply mesheryctl/internal/cli/root/system/error.go:177]: yeah it is for yaml

File: mesheryctl/internal/cli/root/system/error.go
Comment (zakisk) [reply mesheryctl/internal/cli/root/system/error.go:173]: @leecalcote hey, you're right none of these error code are defined by me they were already. and I used them in the same way as they were used by others previously. I should see their usage and suggested remediation :100:

**Comment by zakisk** [comment]
@abdullah1308 as you can see in this PR that I've deleted mesheryctl/internal/cli/root/error.go file because I wanted to refernce them in mesheryctl/internal/cli/root/system package I was getting import cycle error and I was unable to reference them, and error functions in mesheryctl/internal/cli/root/error.go were only used by mesheryctl/internal/cli/root/version.go file so I copied all of them in mesheryctl/internal/cli/root/system/error.go file  and chaged mesheryctl/internal/cli/root/version.go file.

**Comment by abdullah1308** [comment]
@zakisk The approach that me, @Philip-21 and @suhail34 have decided on is to move errors common to all `mesheryctl` commands to `mesheryctl/pkg/utils/error.go`. For errors that are specific to a sub-command they will be stored in the error.go file of sub-command's package.

There's 1/2 PRs up for making this change already. For now I suggest that you undo the change related to deleting the error.go file and moving the error codes. I understand that this causes the import cycle error. However once these other PRs are closed and if master is merged into this PR, these errors should be resolved. Hope this makes sense.

**Comment by zakisk** [comment]
@abdullah1308 error file that I've deleted was used by only one other file i.e version.go so I don't think that it is worth to undo deletion of that file. but if you still suggest me to do so I will add that error file in one more commit to this PR, is that you want?

**Comment by zakisk** [comment]
@abdullah1308 the approach I'm thinking of is that after your PR are merged we can remove all generic errors from mesheryctl in a new PR, does this sound good?

File: mesheryctl/internal/cli/root/system/error.go
Comment (Philip-21) [reply mesheryctl/internal/cli/root/system/error.go:173]: @zakisk nice work here , the prs #8208 and #8256 are being worked on for generating error wrappers to be used in different aspect of `mesheryctl`.  When the pr's are merged you can make changes and call the error functions. so we avoid redundancy in writing error functions in the` mesheryctl/pkg/utils/error.go` just as @abdullah1308 Mentioned

**And please sign off your commits**

**Comment by abdullah1308** [comment]
> @abdullah1308 error file that I've deleted was used by only one other file i.e version.go so I don't think that it is worth to undo deletion of that file. but if you still suggest me to do so I will add that error file in one more commit to this PR, is that you want?

@zakisk Yeah that is what I'd suggest for now.

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [inline mesheryctl/internal/cli/root/system/model.go:68]:
```suggestion
		mctlCfg, err := config.GetMesheryCtl(viper.GetViper())
```
Please Make sure you use similar naming

**Review by suhail34 — CHANGES_REQUESTED** [review]
Can we provide pagination while listing the the model

File: mesheryctl/internal/cli/root/system/error.go
Comment (Philip-21) [inline mesheryctl/internal/cli/root/system/error.go:53]: @zakisk if you look at the pr's #8238 and #8256, most of these error codes and functions you wrote here  are already implemented in the pr's i just listed  , the error functions in the `utils package` are for general use for all the packages in mesheryctl for returning errors,  you may have to remove some of these here to avoid redudancy . 
Only create the ones applicable to the `system` package, or you can add more to the utils package as far as it hasn't been implemented already in the pr's  i mentioned.
please look through the pr's .

**Review by Philip-21 — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/error.go
Comment (zakisk) [reply mesheryctl/internal/cli/root/system/error.go:53]: @Philip-21 as PR #8256 contains error code those I've created in system/error.go but this PR isn't meged yet, am I supposed to wait for its merge? or anything else better you suggest me?

**Comment by zakisk** [comment]
> Can we provide pagination while listing the the model

@suhail34 Would you mind providing further explanation?

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [inline mesheryctl/internal/cli/root/system/model.go:131]:
```suggestion
		header := []string{"Category Name", "Model Name", "Version"}
		rows := [][]string{}

		for _, model := range modelsResponse.Models {
			if len(model.DisplayName) > 0 {
				// in `mesheryctl system model view [model-name]` command `model.Name` must be
				// passed in order to fetch its details
				rows = append(rows, []string{model.Category.Name, model.Name, model.Version})
			}
		}
```
Further more details needed for the user as in the some category there are more than one models with different version

**Review by suhail34 — CHANGES_REQUESTED** [review]

**Comment by suhail34** [comment]
> > Can we provide pagination while listing the the model
> 
> @suhail34 Would you mind providing further explanation?

There are in total 115 model listed after using `mesheryctl system model list` so we need pagination right

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [inline mesheryctl/internal/cli/root/system/model.go:280]:
```suggestion

```
This command aren't of any use so remove the examples

**Review by suhail34 — CHANGES_REQUESTED** [review]

**Comment by zakisk** [comment]
> > > Can we provide pagination while listing the the model
> > 
> > 
> > @suhail34 Would you mind providing further explanation?
> 
> There are in total 115 model listed after using `mesheryctl system model list` so we need pagination right

I mean don't know that how pagination is implemented in CLI?. if there is any other command in mesheryctl that is using pagination point it out here.

File: mesheryctl/internal/cli/root/system/model.go
Comment (zakisk) [reply mesheryctl/internal/cli/root/system/model.go:131]: at the moment we've flow as when user runs `mesheryctl system model view [model-name]` command, if there are more than one models for the given name it shows selector UI (as it is in `mesheryctl login` for provider) for all available versions, if we add version in `model list` command's output then we will have to add a flag in `model view` command for the specification fo version, so in which way do I implement it?.

Comment (zakisk) [reply mesheryctl/internal/cli/root/system/model.go:131]: @suhail34 what I think though about it is that giving selector UI is more convenient and intuitive for user.

**Comment by zakisk** [comment]
@abdullah1308 @Philip-21 I added error.go file please see.

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [reply mesheryctl/internal/cli/root/system/model.go:210]: Make use of this function

**Review by suhail34 — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [inline mesheryctl/internal/cli/root/system/model.go:210]:
```suggestion
		resp, err := utils.MakeRequest(req)
```
Make use of this function

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [reply mesheryctl/internal/cli/root/system/model.go:130]: This is not for view command it is for list command

**Review by suhail34 — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/model.go
Comment (zakisk) [reply mesheryctl/internal/cli/root/system/model.go:210]: @suhail34  ok

File: mesheryctl/internal/cli/root/system/error.go
Comment (Philip-21) [reply mesheryctl/internal/cli/root/system/error.go:53]: you are supposed to wait for both pr's to merge

Comment (Philip-21) [inline mesheryctl/internal/cli/root/system/error.go:53]: you are supposed to wait for both pr's to merge, once theyre are merged you'll update your branch and call the error functions for returning errors . You still have to remove these functions @suhail34 pr and my  pr has some of them already implemented, its why i said u should look through the pr's ,compare them and remove what you have here to avoid duplicacy,  just ` return err` until the branches are merged . Then you can do the importing for returning the error functions/wrappers from the utils package. 
The ones that we didnt implement in our pr's you can define the and use them . 

Is this clear ?

Comment (zakisk) [reply mesheryctl/internal/cli/root/system/error.go:53]: @Philip-21 Yeah, this make much sense.

Comment (zakisk) [reply mesheryctl/internal/cli/root/system/error.go]: @Philip-21 @abdullah1308 I made changes, I think this is what you both want me to do. please check

File: mesheryctl/internal/cli/root/system/model.go
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/system/model.go:270]: @zakisk Will you update this to represent actual behaviour?

Comment (zakisk) [reply mesheryctl/internal/cli/root/system/model.go:270]: @MUzairS15 sorry, I didn't get you, what do you mean by actual behaviour?

Comment (MUzairS15) [reply mesheryctl/internal/cli/root/system/model.go:270]: Above https://github.com/meshery/meshery/pull/8284/files/f44ef2812ad149efd28314cbb056cf7e10040fbe#diff-934d205e38e560ee0c4bcf41dec8cad7b4fae40a75fc034348bd5b6815953d0cR263 it says generates and update components, but comment says `To view list of components`

**Review by Philip-21 — APPROVED** [review]
yh its fine , LGTM

**Comment by vishalvivekm** [comment]
@zakisk Let's discuss this on Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [reply mesheryctl/internal/cli/root/system/model.go]: @leecalcote any suggestion here `mesheryctl system model list` will be listing just model name that are some time repeating because of different version don't you think it should be showing category , model name and version that would give more insights

**Comment by leecalcote** [comment]
Some final feedback to address here, @zakisk

**Comment by zakisk** [comment]
@leecalcote I've done all the changes that were requested.

**Comment by zakisk** [comment]
@leecalcote it is ready to merge.

**Comment by zakisk** [comment]
@leecalcote it is ready to be merged

**Comment by suhail34** [comment]
https://github.com/meshery/meshery/pull/8284#discussion_r1278289599

**Comment by zakisk** [comment]
@suhail34 it wasn't confirmed, [see this](https://github.com/meshery/meshery/pull/8284#discussion_r1285112007), @leecalcote do I change as @suhail34 suggested?

File: mesheryctl/internal/cli/root/system/model.go
Comment (suhail34) [inline mesheryctl/internal/cli/root/system/model.go:192]:
```suggestion
		const errMsg = "Usage: mesheryctl system model view [model-name]\nRun 'mesheryctl system model list' to see list of models"
```

Comment (suhail34) [reply mesheryctl/internal/cli/root/system/model.go:192]: Just this change and every thing LGTM

**Review by suhail34 — CHANGES_REQUESTED** [review]

**Review by suhail34 — APPROVED** [review]
LGTM

**Comment by zakisk** [comment]
@leecalcote

File: mesheryctl/internal/cli/root/system/fixtures/token.golden
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/fixtures/token.golden:1]: Why are we exposing a secret publicly like this?

File: mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden:1]: This will change overtime - potentially daily. Do not hardcode this.

File: mesheryctl/internal/cli/root/system/model.go
Comment (leecalcote) [inline mesheryctl/internal/cli/root/system/model.go:192]: One line of usage for two subcommands isn't enough documentation or example, @zakisk

**Review by leecalcote — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/system/model.go
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/model.go:279]: @zakisk are you going to accept this feedback and incorporate a change?

**Review by suhail34 — CHANGES_REQUESTED** [review]
Get your DCO covered

File: mesheryctl/internal/cli/root/system/model.go
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/model.go]: Still not answered...

File: mesheryctl/internal/cli/root/system/model.go
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/model.go:279]: Be sure that you respond to comments with your changes as you resolve comments, so that it is clear what updates were made, @zakisk

File: mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden
Comment (leecalcote) [reply mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden:1]: Still not answered...

**Comment by suhail34** [comment]
> Get your DCO covered

DCO ....

**Comment by zakisk** [comment]
@alphaX86 @suhail34 fiest five commits aren't signed-off and as I've merged changes from others during the course now my commits are very far to be pointed by `HEAD~n` that's why I've added sign off message in PR's description.

**Comment by Freedisch** [comment]
> @alphaX86 @suhail34 fiest five commits aren't signed-off and as I've merged changes from others during the course now my commits are very far to be pointed by `HEAD~n` that's why I've added sign off message in PR's description.

@zakisk U can sign off specifc commit with `git commit --amend -s -c <SHA>`

**Comment by zakisk** [comment]
@Freedisch the command you've suggested is adding new commit with the same message of old commit of which SHA key is passed it is not amending it on place.

**Comment by alphaX86** [comment]
> @Freedisch the command you've suggested is adding new commit with the same message of old commit of which SHA key is passed it is not amending it on place.

If that's the case, then ignore the DCO for now. We'll take care of that later

File: mesheryctl/internal/cli/root/system/fixtures/token.golden
Comment (alphaX86) [inline mesheryctl/internal/cli/root/system/fixtures/token.golden:1]: @zakisk I recommend NOT to expose the token publicly like THIS! Please remove that or tell the reason on why this is present

File: mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden
Comment (alphaX86) [reply mesheryctl/internal/cli/root/system/testdata/model/list.model.output.golden:1]: You might need explanation for doing this here as this will be dynamic and change from time to time

**Review by alphaX86 — CHANGES_REQUESTED** [review]

**Comment by alphaX86** [comment]
> @alphaX86 API response is stored in golden file and the output will be same every time when tests run, in order to keep uniformity in tests pattern I did this because all other tests you can check in fliter or other commands are following this pattern,

OK, you indeed need a token for them. I understand the requirement. But did you hardcode it?

**Comment by zakisk** [comment]
@alphaX86 yes

**Comment by alphaX86** [comment]
> @alphaX86 yes

Is this token related to your Meshery Cloud account?

**Comment by zakisk** [comment]
@alphaX86 I've copied this from [filter test](https://github.com/meshery/meshery/blob/master/mesheryctl/internal/cli/root/filter/fixtures/token.golden) its not mine.

**Comment by zakisk** [comment]
@alphaX86 and its more than two years old token.

**Comment by alphaX86** [comment]
> @alphaX86 I've copied this from [filter test](https://github.com/meshery/meshery/blob/master/mesheryctl/internal/cli/root/filter/fixtures/token.golden?rgh-link-date=2023-09-09T07%3A49%3A06Z) its not mine.

Ok, noted 👍

**Comment by zakisk** [comment]
@alphaX86 @suhail34 @leecalcote @MUzairS15 is there anything else?

**Comment by zakisk** [comment]
@leecalcote @alphaX86 ??

**Comment by alphaX86** [comment]
> @leecalcote @alphaX86 ??

I'll review this once I'm in my system

**Comment by zakisk** [comment]
> > @leecalcote @alphaX86 ??
> 
> I'll review this once I'm in my system

@alphaX86 have you done this?

**Comment by zakisk** [comment]
@leecalcote @alphaX86 it's been a while this PR is still open.

**Comment by leecalcote** [comment]
@MUzairS15, please review.

**Comment by leecalcote** [comment]
@zakisk great demo today. Thanks for taking everyone through your work. We’re all green 🚦 here.

**Pull request merged**
